Add WP-CLI command to reprocess inbox items

obenland opened the pull request:
## Proposed changes:
* Add `wp activitypub reprocess_inbox` command for local development
* Delete existing comments/posts before reprocessing to ensure clean state
* Enable `activitypub_create_posts` option in local development for testing

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:

1. Set up local development environment with WP-CLI
2. Process an ActivityPub activity to create an inbox item
3. Run `wp activitypub reprocess_inbox <post_id>` where `<post_id>` is an ap_inbox post ID
4. Verify that:
   - Existing comments/posts are deleted
   - The activity is reprocessed
   - New comments/posts are created with the original timestamps

**Example usage:**
```bash
# Reprocess inbox item with ID 123
wp activitypub reprocess_inbox 123
```

**Expected output:**
```
Reprocessing inbox item 123...
Deleted comment 456 (source_id: object ID)
Deleted 1 comment(s) and 0 post(s).
Success: Inbox item 123 has been reprocessed.
```

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

obenland committed:
Add WP-CLI command to reprocess inbox items

Adds 'wp activitypub reprocess_inbox' command for local development that:
- Accepts an ap_inbox post ID
- Deletes any existing comments or ap_post items created from the activity
- Reprocesses the activity through the inbox handler

This is useful for:
- Testing inbox processing logic
- Debugging activity handling
- Re-importing activities after code changes

Also enables activitypub_create_posts option in local development to allow
testing of post creation from ActivityPub Create activities.

Usage: wp activitypub reprocess_inbox <post_id>

Example:
  wp activitypub reprocess_inbox 123

obenland commented on local/class-cli.php:
```diff
@@ -168,4 +171,87 @@ private function generate_comments( $args, $assoc_args ) {
+				\wp_delete_post( $ap_post->ID, true );
+				++$deleted_posts;
+				\WP_CLI::log( sprintf( 'Deleted ap_post %d (GUID: object ID)', $ap_post->ID ) );
+			}
```
At that point it should create a fatal error so we can fix it.

obenland committed:
Address code review feedback: use distinct variable name

Use object_comment instead of reusing comment variable for the second lookup to improve code clarity and prevent potential confusion.

obenland commented on local/class-cli.php:
```diff
@@ -168,4 +171,87 @@ private function generate_comments( $args, $assoc_args ) {
+		}
+
+		// Call the process_create_item method to reprocess.
+		Actors_Inbox_Controller::process_create_item( $activity_id );
```
This is on purpose. We don't want to skip on a duplicated item.

obenland commented on local/class-cli.php:
```diff
@@ -168,4 +171,87 @@ private function generate_comments( $args, $assoc_args ) {
+			if ( $comment ) {
+				\wp_delete_comment( $comment->comment_ID, true );
+				++$deleted_comments;
+				\WP_CLI::log( sprintf( 'Deleted comment %d (source_id: object ID)', $comment->comment_ID ) );
```
Fixed in [f390ee1](https://github.com/Automattic/wordpress-activitypub/pull/2500/commits/f390ee1315d636f19fa9267a6a489d80764178ea).

pfefferle reviewed (changes requested):
I would assume that a method called `reprocess` would reprocess any kind of Inbox Item, not only `Create`s (aside from that, the current implementation would handle any kind of Item as a `Create`). The Inbox stores every incoming Activity (except `Delete`s), so a reprocess has to take care of properly reprocessing `Announce`s, `Like`s, `Undo`s and all other types, the plugin supports.

To have a proper reprocess we should simply call the same `do_action` that is triggered by the inbox, to run through the same detection and handler process, a normal Activity would run through.

pfefferle commented on local/class-cli.php:
```diff
@@ -168,4 +171,87 @@ private function generate_comments( $args, $assoc_args ) {
+		$deleted_posts    = 0;
+
+		// Delete comments with source_id matching the activity ID.
+		$comment = Comment::object_id_to_comment( $activity_id );
```
There is no check if the Activity is a `Create` or `Update`!

obenland replied:
Fixed in [531d041](https://github.com/Automattic/wordpress-activitypub/pull/2500/commits/531d0418fa7785a728004637f9ed9ebcbc1720ab)

obenland committed:
Reprocess inbox items via internal REST API call

Instead of directly calling process_create_item(), the reprocess_inbox
command now makes an internal REST API call to the shared inbox endpoint.
This ensures the activity goes through the full processing pipeline
including validation, moderation, recipient determination, and all
registered action hooks for any activity type.

obenland committed:
Only delete artifacts for Create/Like/Announce activities

Changed reprocess_inbox to conditionally delete existing artifacts based
on activity type. Only Create, Like, and Announce activities now trigger
deletion. This preserves existing posts/comments for Update, Delete, and
Undo activities, which need to find and operate on those artifacts.

obenland committed:
Merge branch 'trunk' into add/reprocess-inbox-cli-command

pfefferle commented on local/class-cli.php:
```diff
@@ -168,4 +170,110 @@ private function generate_comments( $args, $assoc_args ) {
+		$deleted_comments = 0;
+		$deleted_posts    = 0;
+
+		if ( $should_delete ) {
```
why deleting existing comments/posts? the plugin should be able to handle these cases automatically.

obenland replied:
It doesn't seem to. When reprocessing create activities, ap_posts don't get updated. Not sure if this is a bug, but for now deleting does what I need this command to do.

pfefferle commented on local/class-cli.php:
```diff
@@ -168,4 +170,110 @@ private function generate_comments( $args, $assoc_args ) {
+		\add_filter( 'activitypub_defer_signature_verification', '__return_true' );
+
+		// Create internal REST request to the shared inbox endpoint.
+		$request = new \WP_REST_Request( 'POST', '/' . \ACTIVITYPUB_REST_NAMESPACE . '/inbox' );
```
I am not sure if this is always possible. Some Activities could have been sent to the Actors inbox with an empty audience. In these cases, the Activity wouldn't be reprocessed from the shared-inbox.

obenland replied:
Fixed in [80f4b6a](https://github.com/Automattic/wordpress-activitypub/pull/2500/commits/80f4b6a753fbc7ff75de8650a07e790a22c279e8).

obenland committed:
Reprocess inbox items by calling action hooks directly

The reprocess_inbox command now:
- Deletes existing artifacts before reprocessing (handlers check for
  duplicates and skip creation if they exist)
- Calls both activitypub_inbox_{type} AND activitypub_handled_inbox_{type}
  hooks since handlers are registered on different hook sets
- Preserves original recipients from inbox post meta
- Works for all activity types (Create, Update, Delete, Like, Announce, Undo)

This ensures activities go through the same detection and handler process
as new activities, addressing feedback about proper reprocessing.

obenland commented:
@pfefferle Thanks for your feedback, I think this is in a better shape than it was last night.

This is about as much time I'd want to invest in this PR. If there are any additional points of feedback, I think we're quickly approaching a negative return for time invested, and I'd want to find a different approach.

obenland committed:
Fix lint

pfefferle commented:
> ...and I'd want to find a different approach.

What is the purpose of the PR?

obenland commented:
To allow reprocessing of inbox items in local development to test changes in ap_post parsing. Test if changes like in #2499 are working.

pfefferle commented:
> This is about as much time I'd want to invest in this PR. If there are any additional points of feedback, I think we're quickly approaching a negative return for time invested, and I'd want to find a different approach.

I’m not so sure about that then. I think the time investment is still worthwhile! If the reprocessing isn’t working correctly, then an `Update` Activity from a remote server won’t work properly either.

pfefferle committed:
Merge branch 'trunk' into add/reprocess-inbox-cli-command

pfefferle commented:
@obenland can you provide me some examples to test why the reprocess only work with the deletion?

obenland commented:
The way I tested it was to change an ap_post item, remove some content for example, then reprocess its inbox item and see if the content was restored.

pfefferle commented:
@obenland found the issue! will work on a fix!

pfefferle committed:
Refactor activity handling to use direct method calls

Replaces do_action hooks with direct calls to Update::handle_update and Create::handle_create for improved clarity and flow in activity handling. Removes artifact deletion logic from CLI reprocessing, streamlining the process.

pfefferle commented:
@obenland should be fixed now!

obenland commented:
@pfefferle Yes, looks good. Has some unit tests failing now.

You'll have to accept the PR when you're happy with it ;)

pfefferle committed:
Update test to use activitypub_handled_create action

Replaces references to the deprecated 'activitypub_inbox_create' action with 'activitypub_handled_create' in the test. Also updates the test activity object and adjusts the action handler parameters for compatibility.

pfefferle commented:
Now that I’ve seen the effects of the restructuring, I’m not sure “happy with it” is the right way to put it. :(
